src/transformers/models/gpt_neox/modeling_flax_gpt_neox.py

```diff
@@ -0,0 +1,732 @@
+
+    sentinel = dim // 2 + dim % 2
+    out = np.zeros((num_pos, dim))
+    out[:, 0:sentinel] = sin
+    out[:, sentinel:] = cos
+
+    return jnp.array(out)
+
+
+def rotate_half(x):
+    x1 = x[..., : x.shape[-1] // 2]
+    x2 = x[..., x.shape[-1] // 2 :]
+    return jnp.concatenate((-x2, x1), axis=-1)
+
+
+class FlaxGPTNeoXRotaryEmbedding(nn.Module):
+    dim: int
+    max_seq_len_cached: int
+    base: int = 10000
+
+    def setup(self):
+        self.inv_freq = 1.0 / (self.base ** (jnp.arange(0, self.dim, 2).astype(jnp.float32) / self.dim))  # dim
+        self._init_cache(self.max_seq_len_cached)
+
+    def _init_cache(self, seq_len: int):
+        t = jnp.arange(seq_len, dtype=self.inv_freq.dtype)
+        freqs = jnp.outer(t, self.inv_freq)
+        emb = jnp.concatenate((freqs, freqs), axis=-1)
+        self.cos_cached = jnp.expand_dims(jnp.expand_dims(jnp.cos(emb), 0), 0)
+        self.sin_cached = jnp.expand_dims(jnp.expand_dims(jnp.sin(emb), 0), 0)
+
+    def __call__(self, seq_len=None):
+        if seq_len > self.max_seq_len_cached:
+            self._init_cache(seq_len)
+            self.max_seq_len_cached = seq_len
+
+        return self.cos_cached[:seq_len, ...], self.sin_cached[:seq_len, ...]
+
+
+def apply_rotary_pos_emb(q, k, cos, sin, position_ids):
+    gather_indices = position_ids[:, :, None, None]  # [bs, seq_len, 1, 1]
+    gather_indices = jnp.repeat(gather_indices, cos.shape[1], axis=1)
+    gather_indices = jnp.repeat(gather_indices, cos.shape[3], axis=3)
+    cos = jnp.take_along_axis(cos.repeat(gather_indices.shape[0], axis=0), gather_indices, axis=2)
+    sin = jnp.take_along_axis(sin.repeat(gather_indices.shape[0], axis=0), gather_indices, axis=2)
+    q_embed = (q * cos) + (rotate_half(q) * sin)
+    k_embed = (k * cos) + (rotate_half(k) * sin)
+    return q_embed, k_embed
+
+
+class FlaxGPTNeoXAttention(nn.Module):
+    config: GPTNeoXConfig
+    dtype: jnp.dtype = jnp.float32
+
+    def setup(self):
+        config = self.config
+        self.num_attention_heads = config.num_attention_heads
+        self.hidden_size = config.hidden_size
+        self.head_size = self.hidden_size // self.num_attention_heads
+        self.rotary_ndims = int(self.head_size * config.rotary_pct)
+        self.embed_positions = create_sinusoidal_positions(config.max_position_embeddings, self.rotary_ndims)
+        self.norm_factor = jnp.sqrt(self.head_size)
+        self.query_key_value = nn.Dense(
+            3 * config.hidden_size,
+            dtype=self.dtype,
+            kernel_init=jax.nn.initializers.normal(self.config.initializer_range),
+        )
+        self.dense = nn.Dense(
+            config.hidden_size,
+            dtype=self.dtype,
+            kernel_init=jax.nn.initializers.normal(self.config.initializer_range),
+        )
+
+        self.causal_mask = make_causal_mask(jnp.ones((1, config.max_position_embeddings), dtype="bool"), dtype="bool")
+
+        self.rotary_emb = FlaxGPTNeoXRotaryEmbedding(
```

Comment:
I copied RoPE scaling code and test code. But in [test CI](https://app.circleci.com/pipelines/github/huggingface/transformers/74067/workflows/e61221ea-2282-4e69-ae4c-1fa4b0e28f30/jobs/937874), the frozen flax model raises SetAttributeFrozenModuleError.

For the cached RoPE embedding, I think I should use the variable. I'm trying to implement it, and I think it'd be nice if you had an appropriate reference or suggestion. @sanchit-gandhi

Reply:
Ah we can only set attributes in Flax in the `setup` method. After that, the module gets frozen, meaning we can't add new attributes or update existing ones. So probably the cached embedding isn't going to work - what we can instead do is always initialise the embeddings to max length (`config.max_position_embeddings`), and then slice the first `N` embeddings as required each time. This way, we don't ever need to re-compute or update the embeddings, since we always have the max embedding length we require stored in the `setup`